packages/types/mongoose-store/index.ts

```diff
@@ -0,0 +1,64 @@
+import * as Promise from 'bluebird';
+import * as mongoose from 'mongoose';
+
+export default class MongooseStore<T>{
+  constructor(protected conn: mongoose.Connection) {
+
+  }
+}
+
+type ID = string | number;
+
+export interface Store<T> {
+  query(): ExecutableChainQuery<T>;
+  or(): ChainQuery<T>;
+  find(query: QueryParams<T>): Promise<T[]>;
+  create(data: T): Promise<T>;
+  update(id: ID, data: T): Promise<T>;
+  delete(id: ID): Promise<T>;
+}
+
+interface Predicate<T> {
```

Comment:
I thought our requirements for the storage module involved:

- Having a public API that could be implemented with a different storage back-end
- Having as much compile-time type checking as possible
- Support POJOs as models

I know it's a big reinvented wheel, but at least I don't think that there's any ORM for node out there that works well with TS.
If we're to build our own, I'd want use the base `mongodb` driver module rather than `mongoose`.

Apart from that, if we start using mongoose in our other modules it's going to be hard to refactor out of it, which can be okay, of course. This gets even harder as we start using more specific features like schema-based validation and subdocuments, and so on. That's why a non-storage specific module would need to do data validation mostly on the nodejs side.

It would be possible for our users to use features specific to their storage engine choice on their own code, while they'd be required to write their own implementation of the public API so it'd work with our own modules, those could also be contributed back as an OSS, upstream-supported module.